Add bnpl logos to thank you page.

brettshumaker opened the pull request:
Fixes #8249

#### Changes proposed in this Pull Request
This PR adds the bnpl payment method logo to the thank you page. I'm utilizing the `get_theme_icon()` to get the light/dark icon based on whichever checkout upe transient has been set (classic or block checkout). WooPay also adds a logo in this spot so I borrowed the approach used there but then generalized the wrapper class so I didn't have to repeat CSS. I also added a WooPay-specific class to that wrapper as well as the new bnpl wrappers in case we ever need to scope CSS to them individually. This implementation also adds a `wc_payments_thank_you_page_bnpl_payment_method_logo_url` filter to allow users to change the logo URL in case our transient specifies the dark logo but a light logo works better on the thank you page. It also falls back to the payment method title (how it currently works) if it can't get the logo URL for some reason.


#### Testing instructions
* Complete a checkout with a BNPL method and you should see the logo on the thank you page instead of the plain title.
* Please check multiple themes with both light and dark versions to ensure that the proper logo style is pulled. You'll need to visit the checkout page after the theme change to make sure that the UPE appearances are generated and saved.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

brettshumaker committed:
Add bnpl logo to thank you page.

* Adds the bnpl logo to the thank you page.
* Generalizes and reuses CSS class for both WooPay and BNPL
* Add max-height to bnpl logos

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8650` or branch name `8249/add-bnpl-icons-to-thank-you-page` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=8249/add-bnpl-icons-to-thank-you-page) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 9f92c521065cf2bec36bc98de9d20bbb725b85fb
- Build time: 2024-04-30 10:22:52 UTC

_Note: the build is updated when a new commit is pushed to this PR._

brettshumaker committed:
If no method logo url, return the method title

brettshumaker committed:
Merge branch 'develop' into 8249/add-bnpl-icons-to-thank-you-page

brettshumaker committed:
Added `woopay` class to its wrapper and updated php tests

brettshumaker committed:
Merge branch '8249/add-bnpl-icons-to-thank-you-page' of https://github.com/Automattic/woocommerce-payments into 8249/add-bnpl-icons-to-thank-you-page

brettshumaker committed:
Merge branch 'develop' into 8249/add-bnpl-icons-to-thank-you-page

brettshumaker committed:
Add bnpl method id to wrapper class

brettshumaker committed:
Add `wc_payments_thank_you_page_bnpl_payment_method_logo_url` filter.

This is to allow users to change the logo URL if needed. A specific use case is if we've returned the `dark_icon` URL but the regular icon is preferred/required.

brettshumaker committed:
Create 8249-add-bnpl-icons-to-thank-you-page

FangedParakeet commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -4253,10 +4253,13 @@ public function wc_payments_get_payment_method_by_id( $payment_method_id ) {
-		$upe_appearance_theme = get_transient( self::UPE_APPEARANCE_THEME_TRANSIENT );
-		if ( $upe_appearance_theme ) {
-			return 'night' === $upe_appearance_theme ? $this->payment_method->get_dark_icon() : $this->payment_method->get_icon();
+		$upe_appearance_theme        = get_transient( self::UPE_APPEARANCE_THEME_TRANSIENT );
+		$upe_blocks_appearance_theme = get_transient( self::WC_BLOCKS_UPE_APPEARANCE_THEME_TRANSIENT );
+
+		if ( 'night' === $upe_appearance_theme || 'night' === $upe_blocks_appearance_theme ) {
```
Just FYI, this `get_theme_icon()` function originally does not use the blocks theme transient, because this function is only used to decide the payment method icon to display in the shortcode checkout [here](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payment-gateway-wcpay.php#L280). The shortcode checkout logo for payment methods relies on the `icon` property of the payment gateway, whereas the logo for blocks checkout is [inserted into the DOM manually](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/blocks/payment-method-label.js#L59-L67).

I think it's unlikely that shortcode and blocks checkout will require opposite themes, but it's possible and would create an unintended side-effect here. This `get_theme_icon()` function can probably be named more appropriately (ma b, I believe) to express its actual purpose, but I'd probably recommend persisting its original logic. 🤷

FangedParakeet commented on includes/class-wc-payments-order-success-page.php:
```diff
@@ -73,6 +75,58 @@ public function show_woopay_payment_method_name( $payment_method_title, $abstrac
+
+		$order_id = $abstract_order->get_id();
+		$order    = wc_get_order( $order_id );
+		$bnpl_methods = [
+			'Affirm',
+			'Afterpay',
+			'Clearpay',
+			'Klarna',
+		];
+
+		if ( ! $order || ! in_array( $payment_method_title, $bnpl_methods, true ) ) {
+			return $payment_method_title;
+		}
```
```suggestion
		$bnpl_methods = [
			new Affirm_Payment_Method()->get_title(),
			new Afterpay_Payment_Method()->get_title(),
			new Clearpay_Payment_Method()->get_title(),
			new Klarna_Payment_Method()->get_title(),
		];

		if ( ! $order || ! in_array( $payment_method_title, $bnpl_methods, true ) ) {
			return $payment_method_title;
		}
```
This comparison here is a little rigid and has potential to crack. Can we use the title from the `UPE_Payment_Method` class instead? Note this example is just one example of how to gather these titles and might not even be the best one--you can probably get these instantiated classes from one of the static WCPay map utility functions too.

FangedParakeet commented on includes/class-wc-payments-order-success-page.php:
```diff
@@ -73,6 +75,58 @@ public function show_woopay_payment_method_name( $payment_method_title, $abstrac
+	 * @param string            $payment_method_title the default payment method title.
+	 * @param WC_Abstract_Order $abstract_order the order being shown.
+	 */
+	public function show_bnpl_payment_method_name( $payment_method_title, $abstract_order ) {
```
This function shares a lot of code with the `show_woopay_payment_method_name()` function: did you consider refactoring to combine them or did it not appear practical? Just thinking out aloud.

FangedParakeet commented on includes/class-wc-payments-order-success-page.php:
```diff
@@ -73,6 +75,58 @@ public function show_woopay_payment_method_name( $payment_method_title, $abstrac
+
+		$order_id = $abstract_order->get_id();
+		$order    = wc_get_order( $order_id );
+		$bnpl_methods = [
+			'Affirm',
+			'Afterpay',
+			'Clearpay',
+			'Klarna',
+		];
+
+		if ( ! $order || ! in_array( $payment_method_title, $bnpl_methods, true ) ) {
+			return $payment_method_title;
+		}
+
+		$payment_method_id = $order->get_payment_method();
+		$payment_method    = WC()->payment_gateways()->payment_gateways()[ $payment_method_id ];
+
+		if ( ! $payment_method ) {
+			return $payment_method_title;
+		}
```
```suggestion
		$payment_method_id = $order->get_payment_method();
		$payment_method = WC_Payments::get_payment_method_by_id( $payment_method_id )

		if ( ! $payment_method ) {
			return $payment_method_title;
		}
```
Couple points here. Do we need to restrict this functionality to BNPL payment methods? BNPL methods were mentioned as examples on the original issue, but I didn't read this as a restriction. Moreover, the rest of the code begins significantly simpler if we just remove this restriction and allow all payment method icons to be displayed, since you've already made the rest of this code dynamic.

FangedParakeet replied:
```suggestion
		$bnpl_methods = [
			'woocommerce_payments_' . Affirm_Payment_Method::PAYMENT_METHOD_STRIPE_ID,
			...
		];

		$payment_method_id = $order->get_payment_method();
		if ( ! $order || ! in_array( $payment_method_id, $bnpl_methods, true ) ) {
			return $payment_method_title;
		}

		$payment_method    = WC()->payment_gateways()->payment_gateways()[ $payment_method_id ];

		if ( ! $payment_method ) {
			return $payment_method_title;
		}
```
In any case, even if we are restricting the available payment methods here, we should be using the class constants to ensure this check on the title is not so brittle.

FangedParakeet commented:
I've left a few comments, but I'm to regretfully reassign my review, since I won't have time to give this a second look this afternoon, will be travelling tomorrow, and do not want to be a blocker for this PR. 🙏

brettshumaker committed:
Merge branch 'develop' into 8249/add-bnpl-icons-to-thank-you-page

brettshumaker committed:
Add is_bnpl property to UPE payment method classes.

brettshumaker committed:
Refactor methods hooked into woocommerce_order_get_payment_method_title to share some functionality

brettshumaker committed:
Update woopay order success page tests to account for new methods

brettshumaker committed:
Create a new method to get the payment method icon from a payment method for a given location and context.

brettshumaker committed:
Merge branch 'develop' into 8249/add-bnpl-icons-to-thank-you-page

brettshumaker committed:
Properly revert my changes to get_theme_icon()

brettshumaker committed:
Appease the php linter

brettshumaker committed:
Merge branch 'develop' into 8249/add-bnpl-icons-to-thank-you-page

FangedParakeet commented on includes/class-wc-payments-order-success-page.php:
```diff
@@ -28,39 +28,74 @@ public function __construct() {
+			return $this->show_woopay_payment_method_name( $order );
+		}
+
+		$gateway           = WC()->payment_gateways()->payment_gateways()[ $payment_method_id ];
+		$payment_method    = $gateway->get_payment_method( $order );
+
+		// If this is a BNPL order, return the html for the BNPL payment method name.
+		if ( $payment_method->is_bnpl() ) {
+			$bnpl_output = $this->show_bnpl_payment_method_name( $gateway, $payment_method );
+
+			if ( $bnpl_output !== false ) {
+				return $bnpl_output;
+			}
+		}
```
```suggestion
		// If this is a BNPL order, return the html for the BNPL payment method name.
		if ( $payment_method->is_bnpl() ) {
			$gateway                = WC()->payment_gateways()->payment_gateways()[ $payment_method_id ];
			$payment_method = $gateway->get_payment_method( $order );
			$bnpl_output          = $this->show_bnpl_payment_method_name( $gateway, $payment_method );

			if ( $bnpl_output !== false ) {
				return $bnpl_output;
			}
		}
```
I think we can move these two lines inside the conditional here. We only need the gateway and payment method if the order is a BNPL order AFAICT.

FangedParakeet replied:
Disregard, no we can't lol.

brettshumaker committed:
Make sure the gateway exists and that the get_payment_method method exists before trying to use it.

brettshumaker committed:
Move from `block` to `blocks`

FangedParakeet commented on includes/class-wc-payments-order-success-page.php:
```diff
@@ -28,39 +28,79 @@ public function __construct() {
-	 * @param string            $payment_method_title the default payment method title.
-	 * @param WC_Abstract_Order $abstract_order the order being shown.
+	 * @param string $payment_method_title
+	 * @param WC_Abstract_Order $abstract_order
```
```suggestion
	 * @param string $payment_method_title Original default payment method title.
	 * @param WC_Abstract_Order $abstract_order Successful received order being shown.
```
Not a huge deal, but it's proper documentation form to include a description of the parameters at the end of the doc comment here.

brettshumaker replied:
Resolved in 6744f07a7

FangedParakeet commented on includes/class-wc-payments-order-success-page.php:
```diff
@@ -28,39 +28,79 @@ public function __construct() {
-		$order    = wc_get_order( $order_id );
-		if ( ! $order || ! $order->get_meta( 'is_woopay' ) ) {
+		$order_id          = $abstract_order->get_id();
+		$order             = wc_get_order( $order_id );
+		$payment_method_id = $order->get_payment_method();
+
+		if ( stripos( $payment_method_id, 'woocommerce_payments' ) !== 0 ) {
+			return $payment_method_title;
+		}
+
+		if ( ! $order ) {
+			return $payment_method_title;
+		}
```
```suggestion
		$order             = wc_get_order( $order_id );
		
		if ( ! $order ) {
			return $payment_method_title;
		}
		$payment_method_id = $order->get_payment_method();

		if ( stripos( $payment_method_id, 'woocommerce_payments' ) !== 0 ) {
			return $payment_method_title;
		}
```
I think we should check for the presence of a valid `$order` before we start calling methods on it, otherwise this could produce an error.

brettshumaker replied:
Good call. Resolved in 16ba1cbd3

FangedParakeet commented on includes/class-wc-payments-order-success-page.php:
```diff
@@ -28,39 +28,79 @@ public function __construct() {
+
+		$gateway = WC()->payment_gateways()->payment_gateways()[ $payment_method_id ];
+
+		if ( ! $gateway || ! method_exists( $gateway, 'get_payment_method' ) ) {
```
```suggestion
		if ( ! $gateway ) {
			return $payment_method_title;
		}
		
		if( ! method_exists( $gateway, 'get_payment_method' ) ) {
```
I'd consider separating these conditionals, because--and I'm not entirely sure how this could happen, but nonetheless--if `$gateway` is somehow something truthy, but not an object or a string (e.g. let's say `$gateway` is somehow returned as `1` or `true`), in PHP 8 `method_exists` will throw a fatal error. 

I think triggering this would require someone to inject something foreign into the `payment_gateways()` array, which is farfetched, but not impossible, since ultimately we are relying on a function outside of our plugin at this point. Just worth a thought. 🤷

brettshumaker replied:
Yeah, `$gateway` here could end up being `null` (and throw a warning about an undefined index) if the payment method ID we pass in somehow doesn't exist in the array of gateways we get back.

Resolved in a9792de5b

FangedParakeet commented on includes/class-wc-payments-order-success-page.php:
```diff
@@ -73,6 +113,35 @@ public function show_woopay_payment_method_name( $payment_method_title, $abstrac
+
+		ob_start();
+		?>
+		<div class="wc-payment-gateway-method-logo-wrapper wc-payment-bnpl-logo <?php echo str_replace( 'woocommerce_payments_', '', $payment_method->get_id() ); ?>">
```
```suggestion
		<div class="wc-payment-gateway-method-logo-wrapper wc-payment-bnpl-logo <?php echo $payment_method->get_id(); ?>">
```
If I'm following along correctly, `get_id()` on the `UPE_Payment_Method` class will return the [Stripe ID](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-method.php#L116), not the payment gateway ID, so I think removing this prefix is redundant.

brettshumaker replied:
Yep. You're correct. I think this was left over from a previous iteration where I was using the `payment_method_id` directly from the order and not the payment method.

FangedParakeet commented on includes/class-wc-payments-utils.php:
```diff
@@ -1098,4 +1098,57 @@ public static function is_cart_page(): bool {
+		}
+
+		// Fallback to 'stripe' if no transients are set.
+		return 'stripe';
```
This should probably be a `const` somewhere, but we can leave this for the refactor.

FangedParakeet commented on includes/payment-methods/class-upe-payment-method.php:
```diff
@@ -258,6 +275,24 @@ public function get_dark_icon( string $account_country = null ) {
+	 * @return string
+	 */
+	public function get_payment_method_icon_for_location( string $location = 'checkout', bool $is_blocks = true, string $account_country = null ) {
+		$appearance_theme = WC_Payments_Utils::get_active_upe_theme_transient_for_location( $location, $is_blocks ? 'block' : 'classic' );
```
```suggestion
		$appearance_theme = WC_Payments_Utils::get_active_upe_theme_transient_for_location( $location, $is_blocks ? 'blocks' : 'classic' );
```
I think this should be "blocks" now. (Sorry.)

FangedParakeet commented on includes/payment-methods/class-upe-payment-method.php:
```diff
@@ -258,6 +275,24 @@ public function get_dark_icon( string $account_country = null ) {
+	public function get_payment_method_icon_for_location( string $location = 'checkout', bool $is_blocks = true, string $account_country = null ) {
+		$appearance_theme = WC_Payments_Utils::get_active_upe_theme_transient_for_location( $location, $is_blocks ? 'block' : 'classic' );
+
+		if ( 'night' === $appearance_theme ) {
```
This should also probably be a `const` stored somewhere similar, but you already know the score.

FangedParakeet commented on tests/unit/test-class-wc-payments-utils.php:
```diff
@@ -555,4 +555,53 @@ public function test_get_filtered_error_status_code_with_api_exception() {
 		$this->assertSame( 400, WC_Payments_Utils::get_filtered_error_status_code( new \WCPay\Exceptions\API_Exception( 'Error: Your card was declined.', 'card_declined', 402 ) ) );
 	}
+
+	public function delete_appearance_theme_transients( $transients ) {
+		foreach ( $transients as $location => $contexts ) {
+			foreach ( $contexts as $context => $transient ) {
+				delete_transient( $transient );
+			}
+		}
+	}
+
+	public function set_appearance_theme_transients( $transients ) {
+		foreach ( $transients as $location => $contexts ) {
+			foreach ( $contexts as $context => $transient ) {
+				set_transient( $transient, $location . '_' . $context . '_value', DAY_IN_SECONDS );
+			}
+		}
+	}
```
```suggestion
	private function delete_appearance_theme_transients( $transients ) {
		foreach ( $transients as $location => $contexts ) {
			foreach ( $contexts as $context => $transient ) {
				delete_transient( $transient );
			}
		}
	}

	private function set_appearance_theme_transients( $transients ) {
		foreach ( $transients as $location => $contexts ) {
			foreach ( $contexts as $context => $transient ) {
				set_transient( $transient, $location . '_' . $context . '_value', DAY_IN_SECONDS );
			}
		}
	}
```
Not really a big deal, but seems like better form to keep utility test functions `private` to help differentiate them from the other tests. 🤷

FangedParakeet reviewed (changes requested):
Testing pretty well for me. Nice one! Just have a couple more comments about a few little things I stumbled upon along the way.

brettshumaker committed:
Add parameter descriptions.

brettshumaker committed:
Defensive check against $order

brettshumaker committed:
Defensive check on $gateway to prevent possible fatal error.

brettshumaker committed:
Update tests/unit/test-class-wc-payments-utils.php

Co-authored-by: Samir Merchant <samir.merchant@automattic.com>

brettshumaker committed:
Update a remaining `block` to `blocks`

Co-authored-by: Samir Merchant <samir.merchant@automattic.com>

brettshumaker committed:
Remove extraneous str_replace

brettshumaker committed:
Merge branch 'develop' into 8249/add-bnpl-icons-to-thank-you-page

FangedParakeet commented on includes/class-wc-payments-order-success-page.php:
```diff
@@ -28,39 +28,84 @@ public function __construct() {
+			return $payment_method_title;
+		}
+
+		if ( ! $order ) {
+			return $payment_method_title;
+		}
```
```suggestion
```
Don't need this here twice.

FangedParakeet commented on includes/class-wc-payments-order-success-page.php:
```diff
@@ -28,39 +28,84 @@ public function __construct() {
+			return $payment_method_title;
+		}
+
+		if ( ! $order ) {
+			return $payment_method_title;
+		}
+
```
```suggestion
```
I think we can remove this duplicate check.

brettshumaker replied:
Resolved in 81b506b1c

FangedParakeet commented on includes/class-wc-payments-order-success-page.php:
```diff
@@ -73,6 +118,35 @@ public function show_woopay_payment_method_name( $payment_method_title, $abstrac
+
+		ob_start();
+		?>
+		<div class="wc-payment-gateway-method-logo-wrapper wc-payment-bnpl-logo <?php echo $payment_method->get_id(); ?>">
```
```suggestion
		<div class="wc-payment-gateway-method-logo-wrapper wc-payment-bnpl-logo">
```
On further thought, do we even need this class here or is this for some future use-case? I don't have strong feelings about removing this or leaving it here, just wondering for my own curiosity.

brettshumaker replied:
I added it for future use or in the case that a particular logo needs specific CSS to make it work with whatever theme is in use.

FangedParakeet reviewed (changes requested):
Sorry, I should have gotten to this earlier today as I only have one small suggestion this time! There's a duplicate conditional we should probably remove; the second item, I'll leave at your own discretion. 

This is still testing as expected for me, so if you can have one last look tomorrow morning, hopefully we can merge this in shortly after. 🙏

brettshumaker committed:
Remove duplicate check on $order

brettshumaker committed:
Merge branch 'develop' into 8249/add-bnpl-icons-to-thank-you-page

FangedParakeet reviewed (approved):
LGTM. Thanks for responding to all of my minor suggestions on this PR! 🙌  🚢
